src/box/lua/config/utils/aboard.lua

```diff
@@ -0,0 +1,95 @@
+-- Alert board.
+
+local datetime = require('datetime')
+local log = require('internal.config.utils.log')
+
+local function aboard_set(self, key, alert)
+    assert(alert.type == 'error' or alert.type == 'warn')
+    if alert.type == 'error' then
+        log.error(alert.message)
+    else
+        log.warn(alert.message)
+    end
+    alert.timestamp = datetime.now()
+    self._alerts[key] = alert
```

Comment:
I've added description for the methods and it seems that it should be enough to follow the logic of the module.

```diff
diff --git a/src/box/lua/config/utils/aboard.lua b/src/box/lua/config/utils/aboard.lua
index 5f6d88b0c4..a910468506 100644
--- a/src/box/lua/config/utils/aboard.lua
+++ b/src/box/lua/config/utils/aboard.lua
@@ -3,6 +3,28 @@
 local datetime = require('datetime')
 local log = require('internal.config.utils.log')
 
+-- Set an alert.
+--
+-- The `key` is an unique indentifier of the alert. Attempt to set
+-- a new alert with the same key replaces the existing alert.
+--
+-- The `key` can be used to drop the alert using :drop().
+--
+-- The `alert` is a table of the following structure.
+--
+-- {
+--     type = 'warn' or 'error',
+--     message = <string>,
+-- }
+--
+-- More fields may be placed into the table and all these fields
+-- are returned from `:alerts()`.
+--
+-- The alert board takes an ownership of the `alert` table and
+-- modifies it.
+--
+-- A timestamp is saved on the `:set()` method call to show it in
+-- the `:alerts()` result.
 local function aboard_set(self, key, alert)
     assert(alert.type == 'error' or alert.type == 'warn')
     if alert.type == 'error' then
@@ -14,21 +36,42 @@ local function aboard_set(self, key, alert)
     self._alerts[key] = alert
 end
 
+-- Get an alert or return nil.
 local function aboard_get(self, key)
     return self._alerts[key]
 end
 
+-- Drop an alert pointed by the given `key`.
+--
+-- The function is no-op if the alert doesn't exist.
+--
+-- The `on_drop` callback (see the aboard.new() function) is
+-- called if the alert had exist.
 local function aboard_drop(self, key)
+    if self._alerts[key] == nil then
+        return
+    end
     self._alerts[key] = nil
     if self._on_drop ~= nil then
         self._on_drop(self, key)
     end
 end
 
+-- Drop all the alerts.
+--
+-- The `on_drop` callback is NOT called.
 local function aboard_clean(self)
     self._alerts = {}
 end
 
+-- Serialize the alerts to show them to a user.
+--
+-- The alerts are sorted by a time of the :set() method call and
+-- returned as an array-like table.
+--
+-- The keys passed to the :set() method are NOT shown in the
+-- result of this method: they're considered as internal
+-- identifiers.
 local function aboard_alerts(self)
     -- Don't return alert keys.
     local alerts = {}
@@ -62,6 +105,7 @@ local function aboard_status(self)
     return status
 end
 
+-- Return `true` if there are no alerts, `false` otherwise.
 local function aboard_is_empty(self)
     return next(self._alerts) == nil
 end
@@ -78,6 +122,16 @@ local mt = {
     },
 }
 
+-- Create a new alert board.
+--
+-- The `on_drop` callback is called on the :drop() method call.
+-- It is not called on the :clean() call.
+--
+-- The callback should have the following prototype.
+--
+-- | local function on_drop_cb(aboard, key)
+-- |     <...>
+-- | end
 local function new(opts)
     local on_drop
     if opts ~= nil then
```

As you can see from the diff I also skipped the `on_drop` callback invocation if the alert doesn't exist before the call. It doesn't affect any user visible behavior, but eliminates unnecessary actions in the code.

The diff is applied to the first patch of the series, where the `aboard` module is introduced.

Reply:
> The only thing that bothers me now is that I still can't be sure from this comment/code below, what fields are always present in the `self._alerts`.

AFAIU, if I add a description for the `_alert` field, it solves the problem? It is done.

```diff
diff --git a/src/box/lua/config/utils/aboard.lua b/src/box/lua/config/utils/aboard.lua
index 58bdc8a0ee..6cfbb000e9 100644
--- a/src/box/lua/config/utils/aboard.lua
+++ b/src/box/lua/config/utils/aboard.lua
@@ -167,6 +167,22 @@ local function new(opts)
     end
 
     return setmetatable({
+        -- _alerts is a table of the following structure.
+        --
+        -- {
+        --     [key1] = {
+        --         type = 'warn' or 'error',
+        --         message = <string>,
+        --         timestamp = <datetime object>,
+        --         <..arbitrary fields..>
+        --     },
+        --     [key2] = {
+        --         <...>
+        --     },
+        --     <...>
+        -- }
+        --
+        -- The key can be a string or a number.
         _alerts = {},
         _next_key = 1,
         _on_drop = on_drop,

```

> <...> maybe verify whole argument structure (add `assert(type(alert.message) == "string"`) and <...>

I'm not sure that I should forbid table/cdata/userdata with `__tostring` and/or `__serialize` as a message, so I would leave it unchecked for now.

Reply:
> AFAIU, if I add a description for the _alert field, it solves the problem? It is done.

Thanks, that's great.

> I'm not sure that I should forbid table/cdata/userdata with __tostring and/or __serialize as a message, so I would leave it unchecked for now.

I didn't thing about other types of messages. While I still don't really expect anything else but string here, I agree with your explanation. The only "but" is the comment message, it states message type only as `<string>`. I suggest replacing it with something like `<string> or <serializable>`.

Reply:
> I don't want to mark non-string values as a proper input for the message field. I would better add the assert. I'll do if you insist.

Personally I don't really see a reason for such in-place serialization, and I am not a big fan of a "gray" API. On the other hand, this isn't a blocker for me, so do as you wish.